kntest/pkg/kubetest2/gke/options.go

```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gke
+
+import (
+	"github.com/spf13/cobra"
+
+	"knative.dev/test-infra/pkg/clustermanager/kubetest2"
+)
+
+func addOptions(gkeCmd *cobra.Command, cfg *kubetest2.GKEClusterConfig) {
+	f := gkeCmd.Flags()
+	f.StringVar(&cfg.Environment, "environment", "prod", "The GKE environment, must be one of prod, staging, staging2 and test.")
+	f.StringVar(&cfg.CommandGroup, "command-group", "beta", "The gcloud command group, must be alpha, beta or empty.")
+	f.StringVar(&cfg.GCPProjectID, "gcp-project-id", "", "GCP project ID for creating the cluster")
+	f.StringVar(&cfg.Name, "name", "e2e-cls", "The GKE cluster name.")
+	f.StringVar(&cfg.Region, "region", "us-central1", "The region to create the GKE cluster.")
+	f.StringSliceVar(&cfg.BackupRegions, "backup-regions", []string{"us-west1", "us-east1"}, "The backup regions if the cluster creation runs into stockout issue in the primary region.")
+	f.StringVar(&cfg.Machine, "machine", "e2-standard-4", "The machine type for the GKE cluster.")
+	f.IntVar(&cfg.MinNodes, "min-nodes", 1, "The minimum number of nodes.")
+	f.IntVar(&cfg.MaxNodes, "max-nodes", 3, "The maximum number of nodes.")
+	f.StringVar(&cfg.Network, "network", "e2e-network", "The network name for the GKE cluster.")
+	f.StringVar(&cfg.Version, "version", "latest", "The version of the GKE cluster.")
+	f.StringVar(&cfg.Scopes, "scopes", "cloud-platform", "Scopes for the GKE cluster, should be comma-separated.")
+	f.StringVar(&cfg.Addons, "addons", "", "Addons for the GKE cluster, should be comma-separated.")
+	f.StringVar(&cfg.PrivateClusterAccessLevel, "private-cluster-access-level", "", "Private cluster access level, if not empty, must be one of 'no', 'limited' or 'unrestricted'")
+	f.StringVar(&cfg.PrivateClusterMasterIPRange, "private-cluster-master-ip-range", "172.16.0.%d/28", "The master IP range for the private cluster. The last digit must be left unformatted to allow retrying.")
```

Comment:
This can have some security concerns. Could you split this argument as two: subnet range (172.16.0) and mask (28) or some other combination?
Assuming invokers will provide a %d literal is risky, at minimum.

Reply:
Could you explain a bit why this could cause security issues? I can split the argument as two, but I would like to understand the underlying reason for this.

Reply:
From my understanding on the change, we are requesting to the invoker to provide an IP range value which includes a %d literal in order to fan out the last octet of the address. This is forcing invokers to provide this value in every case or our logic will fail.
This behavior is undesirable, as making this mandatory does not introduce any particular benefit to the customer and it forces every use case to follow that rule.

In terms of security, letting customers provide formatting string literals (and specially one that is expecting particular data types) is not a recommended approach. It is the basis of the string format vulnerability/attack and, from what I've seen, we're not validating the input in any way (_fmt.Sprintf(cc.PrivateClusterMasterIPRange, i))_). I'm unfamiliar with how Go deals with these cases, but it is not a recommended practice overall.

Reply:
Ack, these are very valid reasons. I have changed the flags as suggested.